tests/Unit/Evolution/Systems/GeneralizedHarmonic/BoundaryCorrections/UpwindPenalty.py

```diff
@@ -0,0 +1,197 @@
+                                           mesh_velocity,
+                                           normal_dot_mesh_velocity))
+
+
+def dg_package_data_char_speed_gamma2_v_spacetime_metric(
+    spacetime_metric, pi, phi, constraint_gamma1, constraint_gamma2, lapse,
+    shift, normal_covector, normal_vector, mesh_velocity,
+    normal_dot_mesh_velocity):
+    if normal_dot_mesh_velocity is None:
+        return -(1.0 + constraint_gamma1) * np.dot(
+            shift, normal_covector) * spacetime_metric * constraint_gamma2
+    else:
+        return (
+            -(1.0 + constraint_gamma1) * np.dot(shift, normal_covector) -
+            normal_dot_mesh_velocity) * spacetime_metric * constraint_gamma2
+
+
+def dg_package_data_char_speeds(spacetime_metric, pi, phi, constraint_gamma1,
+                                constraint_gamma2, lapse, shift,
+                                normal_covector, normal_vector, mesh_velocity,
+                                normal_dot_mesh_velocity):
+    result = np.zeros([4])
+    result[0] = -(1.0 + constraint_gamma1) * np.dot(shift, normal_covector)
+    result[1] = -np.dot(shift, normal_covector)
+    result[2] = lapse - np.dot(shift, normal_covector)
+    result[3] = -lapse - np.dot(shift, normal_covector)
+    if not (normal_dot_mesh_velocity is None):
+        for i in range(4):
+            result[i] -= normal_dot_mesh_velocity
+    return result
+
+
+def dg_boundary_terms_spacetime_metric(
+    int_char_speed_v_spacetime_metric, int_char_speed_v_zero,
+    int_char_speed_v_plus, int_char_speed_v_minus,
+    int_char_speed_v_plus_times_normal, int_char_speed_v_minus_times_normal,
+    int_char_speed_gamma2_v_spacetime_metric, int_char_speeds,
+    ext_char_speed_v_spacetime_metric, ext_char_speed_v_zero,
+    ext_char_speed_v_plus, ext_char_speed_v_minus,
+    ext_char_speed_v_plus_times_normal, ext_char_speed_v_minus_times_normal,
+    ext_char_speed_gamma2_v_spacetime_metric, ext_char_speeds,
+    use_strong_form):
+    result = int_char_speed_v_spacetime_metric * 0.
+    if ext_char_speeds[0] > 0.:
+        result -= ext_char_speed_v_spacetime_metric
+    if int_char_speeds[0] < 0.:
+        result -= int_char_speed_v_spacetime_metric
+    return result
+
+
+def dg_boundary_terms_pi(
+    int_char_speed_v_spacetime_metric, int_char_speed_v_zero,
+    int_char_speed_v_plus, int_char_speed_v_minus,
+    int_char_speed_v_plus_times_normal, int_char_speed_v_minus_times_normal,
+    int_char_speed_gamma2_v_spacetime_metric, int_char_speeds,
+    ext_char_speed_v_spacetime_metric, ext_char_speed_v_zero,
+    ext_char_speed_v_plus, ext_char_speed_v_minus,
+    ext_char_speed_v_plus_times_normal, ext_char_speed_v_minus_times_normal,
+    ext_char_speed_gamma2_v_spacetime_metric, ext_char_speeds,
+    use_strong_form):
+    result = int_char_speed_v_spacetime_metric * 0.
+    # Add v^+ terms
+    if ext_char_speeds[2] > 0.:
+        result -= 0.5 * ext_char_speed_v_plus
+    if int_char_speeds[1] < 0.:
```

Comment:
I am confused by the indices on some of these... isn't the [1] entry the v0 speed?